neo/Consensus/PrepareResponse.cs

```diff
@@ -1,12 +1,21 @@
 ﻿using System.IO;
+using Neo.Network.P2P.Payloads;
+using Neo.IO;
 
 namespace Neo.Consensus
 {
     internal class PrepareResponse : ConsensusMessage
     {
-        public byte[] Signature;
+        /// <summary>
+        /// Speaker PrepareRequest Payload
+        /// </summary>
+        public ConsensusPayload PreparePayload;
+        /// <summary>
+        /// Prepare Request, PreparePayload, payload signature
+        /// </summary>
+        public byte[] ResponseSignature;
```

Comment:
Since every `ConsensusPayload` has a signature, do we still need a `ResponseSignature`?

https://github.com/neo-project/neo/blob/831003582b2ada4aff49244e5c88679b8982cc47/neo/Network/P2P/Payloads/ConsensusPayload.cs#L12-L20

Reply:
You want to optimize everything, Erik! aheuaheuahueauheauea
You are right again, I think that we can use this and it is a nice way to save space! Nice catch.

But the logic would be a little different, because nowadays we use the `ResponseSignature` representing a simple way of signing the PrepareRequest. We would need to perform some minnor adjusts.
Again, I personally prefer to merge as it is (because it is quite of well tested) and then we perform this optimization.
But fell free to change anything you want, Erik. It is up to you. If you decide that it is better to do it now count with all of us for testing it.

In this new commits here https://github.com/igormcoelho/neo/tree/optimized-dbft-part-ii we updated it to All Known PrepRequest Signatures.

If you also think that it is worth we can think about merging it now.

Reply:
You don't need to save all the `ResponseSignature`s. You can relay the received `PrepareResponse` message again when needed.

Reply:
In the current code we just relay one signature per time.

In the PART-II PR we are investigating this idea of sharing all `ResponseSignatures`, it is is not merged here yet. I has some advantages in some cases:

* `B` receives `PrepRequest` from `A` and send `Response`
* `C` receives `Response_B` (indirectly gets `PrepRequest_A`) and send `Response`
* `D` only gets `Response_C`  and it will contain all 3 signatures.

But this is a case to be investigated.

Reply:
The logic also applies to the `FinalSignature` of the header, once we receive `OnCommitAgreement` we can save all of them and share them when we relay our commit agreement.
It looks like a small speed up but I think it has potential benefits when the nodes becomes bigger with more nodes, it can help communication considerably.

Reply:
The `ResponseSignature `is the signature of `PrepareRequest `which  is used to ensure that in response phase, CNs are voting for the same PrepareRequest, eventual the same block. I think it might be necessary for that purpose.

Reply:
I don't think it makes sense to perform two duplicate signatures and verifications on the same message.

Reply:
Which one is duplicated?

Reply:
@erikzhang, take a look at these two commits:
https://github.com/igormcoelho/neo/commit/99b3e87a81a67744c3ee6fdc9d120d1cdf8b7cee
https://github.com/igormcoelho/neo/commit/c8cefebe590d0398a035a74064f8d4d716ea9f28

Reply:
You can just send the original `PrepareRequest` or `PrepareResponse` messages. You don't need to get the `Witness` from them.

Reply:
I implemented I draft related to this, check the message in our above (the other thread).

But the point is about sending all know signatures, @erikzhang.
That two commits of Part-II.

Reply:
https://github.com/igormcoelho/neo/pull/5/

Reply:
`ResponseSignatures` and `PreparePayload `are essential for current logic of Regeneration which makes sure other nodes (disconnected or with low network) will be able to move on to commit phase if there are other nodes are already locked in commit phase.  

Without regeneration, it will be difficult to reach consensus with some nodes are locked in commit phase, and other nodes are not in commit phase which will request change view without success. 

I hope my comment makes sense and be helpful.

Reply:
If a node enters the commit phase, it should lock its state and no longer change view. If only 2 nodes enter the locked state, then the other 5 nodes can change view and generate blocks. If there are 3 nodes entering the locked state, then the remaining 4 nodes cannot change view, they must reach a consensus in the current view.

Reply:
A node can only enter commit phase with M valid signatures of the PrepareReq Message, even the Primary.

Reply:
After receiving those `M` valid SignedPaylods, in the worst case `M - f_{failed}  = f_1` will be locked.

The rest `f_{failed} + f_2 = 2f = M` will ask for change_view and the Healthy/NoByzantine nodes (`f_2`) will join the other `f_1` by receiving their OnRegeneration Payload.

Reply:
@erikzhang, can/should I merge https://github.com/igormcoelho/neo/pull/5?